Fix legacy proxy setting handling

dshcherb opened the pull request:
## Please provide the following details to expedite Pull Request review:

----

## Description of change

Auto-population of no-proxy with controller ip addresses caused
proxyupdater to use legacy proxy settings instead of new juju-* proxy
settings.

This change modifies this logic to first check if new juju proxy
settings are present and auto-populate them with controller IPs,
otherwise legacy proxy settings are auto-populated.

Logging is also improved to be more explicit about the type of settings
applied.

## QA steps


```

# 1) replace IPs and ranges for your environment
cat bootstrap-config.yaml 
juju-http-proxy: http://10.10.101.2:8000
juju-https-proxy: http://10.10.101.2:8000
juju-no-proxy: 'localhost,127.0.0.1,10.10.10.0/24,10.10.101.0/24'
logging-config: '<root>=ERROR;unit=TRACE;juju.worker.proxyupdater=TRACE'

# 2) bootstrap a controller with new proxy settings set
juju bootstrap --config bootstrap-config.yaml --bootstrap-constraints mem=2048 vmaas

# 3) validate machine agent log
juju ssh -m controller 0 sudo -i
# the apiserver machine agent log should show
grep proxyupdater /var/log/juju/machine-0.log
2018-07-17 23:43:50 DEBUG juju.worker.proxyupdater proxyupdater.go:165 applying in-process juju proxy settings proxy.Settings{Http:"http://10.10.101.2:8000", Https:"http://10.10.101.2:8000", Ftp:"", NoProxy:"10.10.10.0/24,10.10.101.0/24,10.10.101.46,127.0.0.1,localhost", AutoNoProxy:""}
2018-07-17 23:43:50 DEBUG juju.worker.proxyupdater proxyupdater.go:188 saving new legacy proxy settings proxy.Settings{Http:"", Https:"", Ftp:"", NoProxy:"127.0.0.1,::1,localhost", AutoNoProxy:""}
2018-07-17 23:43:50 TRACE juju.worker.proxyupdater proxyupdater.go:216 setting snap proxy values: proxy.Settings{Http:"", Https:"", Ftp:"", NoProxy:"", AutoNoProxy:""}, "", ""
2018-07-17 23:43:50 DEBUG juju.worker.proxyupdater proxyupdater.go:233 snap core settings [proxy.http= proxy.https= proxy.store=] updated, output: ""

# 4) verify that the proxy server log does not contain any entries from no-proxy hostnames, IPs or subnet ranges
```

## Documentation changes

No changes to CLI or API - this is fix to make this functionality work as documented.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1782236

manadart commented:
!!build!!

dshcherb commented:
@manadart 

I don't have any access to the CI environment to debug the above and not sure my build requests will be authorized to see if this is an intermittent issue.

My local snap builds were successful though and I was able to bootstrap a controller and test this functionality.

SimonRichardson commented:
!!build!!

dshcherb commented:
Pull-request updated, HEAD is now 13475442f227cbed78f9790e3ef86dfa1851643b

SimonRichardson commented:
!!build!!

dshcherb commented:
Sure, will do.

manadart commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[2.4.0] juju-no-proxy variable is not applied due to legacy no-proxy value being auto-set by juju](https://trello.com/c/Nmusp2JB/381-240-juju-no-proxy-variable-is-not-applied-due-to-legacy-no-proxy-value-being-auto-set-by-juju)

dshcherb commented:
Pull-request updated, HEAD is now 737de7413e1becaa19767100f58aaf191c2683f2

dshcherb commented:
Pull-request updated, HEAD is now 4f8fb4855221143b76ef969ac7a57212a599bdc5

dshcherb commented:
@manadart 

Modified one test and added another one.

Am I right to assume that I can use your CI for test runs or is there any documented process to run tests in a private lab (at least on a high-level)? Is running !!build!! only available for core Juju devs?

SimonRichardson commented:
!!build!!

manadart commented:
$$merge$$
